### PR TITLE
Support CompletableFuture in DialogDisplayer

### PR DESCRIPTION
**Pull request by sdedic**
`DialogDisplayer` API allows to wait for dialog's close when using `notify` method but that wasted a thread to wait for the `notify` method to return. And all the goodies that `CompletableFuture` API can't be used, while this is a perfect example of a Future.

So this PR adds 
```
public <T extends NotifyDescriptor> CompletableFuture<T> notifyCompletion(final T descriptor) {
```
that runs similar to `notifyLater`, but returns a Future that the continuation can chain after. There are important differences, hopefully described in the javadoc:
- the value is the `NotifyDescriptor` subclass itself. This is because for example `NotifyDescriptor.InputLine` has additional getters that are typically processed in the completion handler. This allows to pass NotifyDescriptors from one step to another wihtout a need for a variable in a upper scope or member variable that can be accessed from another closure.
- if the UI is cancelled / closed, this future **completes exceptionally** with a `CancellationException` (as if it was `Future.cancel()`ed). This allows the continuation chain to focus on the main job.
- chained processing runs in an **unspecified thread**, possibly the EDT; any long processing must be added using **async** methods of CompletableFuture so EDT is not blocked. Currently, when using and waiting for `notify`, the whole thread is dedicated.

### CODE REVIEW

File: platform/openide.dialogs/src/org/openide/DialogDisplayer.java [hunk 1]
```diff
@@ -98,6 +101,53 @@ public void run() {
...
+     * builtin thread pool.
+     * <div class="nonnormative">
+     * The following snippet is an example of chained dialogs (can be any other processing): {@codesnippet dialogdisplayer-notifyCompletion}
+     * </div>
+     * @param <T> actual subclass of {@link NotifyDescriptor} passed as a parameter.
+     * @param descriptor describes the UI / dialog.
+     * @return the descriptor instance.
+     * @since 7.60
+     */
+    public <T extends NotifyDescriptor> CompletableFuture<T> notifyCompletion(final T descriptor) {
```
Comment (JaroslavTulach) [inline platform/openide.dialogs/src/org/openide/DialogDisplayer.java:131]: I'd call the method `notifyFuture` - it returns a future, right?

**Review by JaroslavTulach — APPROVED** [review]

**Review by dbalek — APPROVED** [review]
Thanks.

**Comment by sdedic** [comment]
According to @entlicher the reason for the failed test `DebugAllBaseTest`  is most probably already in the master, not in this PR. See for example https://app.travis-ci.com/github/apache/netbeans/builds/246681735 which is failing with the same message.
https://issues.apache.org/jira/browse/NETBEANS-6458 was filled for the suspected cause, but not fixed yet.

Merging.

**Comment by sdedic** [comment]
Rebased on latest master; adapted to changes added to LSP DialogDisplayerImpl by #3584

**Review by dbalek — APPROVED** [review]

File: java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/ui/NotifyDescriptorAdapter.java [hunk 2]
```diff
@@ -277,45 +279,122 @@ public Object clientNotify() {
...
+            qpF.thenAccept(selected -> {
                 if (selected == null) {
-                    return NotifyDescriptor.CLOSED_OPTION;
+                    descriptor.setValue(NotifyDescriptor.CLOSED_OPTION);
+                    ctrl.completeExceptionally(new CancellationException());
+                } else {
+                    for (Map.Entry<QuickPickItem, NotifyDescriptor.QuickPick.Item> entry : items.entrySet()) {
+                        entry.getValue().setSelected(selected.contains(entry.getKey()));
+                    }
+                    descriptor.setValue(NotifyDescriptor.OK_OPTION);
```
[the remark this answers is not in the document]
Comment (sdedic) [reply java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/ui/NotifyDescriptorAdapter.java:362]: Recommitted as  5da86ff

**Comment by sdedic** [comment]
All OK again except the GraalVM LTS test, see comments above. Finally merging.

**Pull request merged**